xds/src/main/java/io/grpc/xds/XdsDependencyManager.java

```diff
@@ -0,0 +1,597 @@
+ * maintains the watchers for the xds resources and when an update is received, it either requests
+ * referenced resources or updates the XdsConfig and notifies the XdsConfigWatcher.  Each instance
+ * applies to a single data plane authority.
+ */
+final class XdsDependencyManager implements XdsConfig.XdsClusterSubscriptionRegistry {
+  public static final XdsClusterResource CLUSTER_RESOURCE = XdsClusterResource.getInstance();
+  public static final XdsEndpointResource ENDPOINT_RESOURCE = XdsEndpointResource.getInstance();
+  private final XdsClient xdsClient;
+  private final XdsConfigWatcher xdsConfigWatcher;
+  private final SynchronizationContext syncContext;
+  private final String dataPlaneAuthority;
+  private final Map<String, Set<ClusterSubscription>> clusterSubscriptions = new HashMap<>();
+
+  private final InternalLogId logId;
+  private final XdsLogger logger;
+  private XdsConfig lastXdsConfig = null;
+  private final Map<XdsResourceType<?>, TypeWatchers<?>> resourceWatchers = new HashMap<>();
+
+  XdsDependencyManager(XdsClient xdsClient, XdsConfigWatcher xdsConfigWatcher,
+                              SynchronizationContext syncContext, String dataPlaneAuthority,
+                              String listenerName) {
+    logId = InternalLogId.allocate("xds-dependency-manager", listenerName);
+    logger = XdsLogger.withLogId(logId);
+    this.xdsClient = xdsClient;
+    this.xdsConfigWatcher = xdsConfigWatcher;
+    this.syncContext = syncContext;
+    this.dataPlaneAuthority = checkNotNull(dataPlaneAuthority, "dataPlaneAuthority");
+
+    // start the ball rolling
+    addWatcher(new LdsWatcher(listenerName));
+  }
+
+  @Override
+  public ClusterSubscription subscribeToCluster(String clusterName) {
+
+    checkNotNull(clusterName, "clusterName");
+    ClusterSubscription subscription = new ClusterSubscription(clusterName);
+
+    Set<ClusterSubscription> localSubscriptions =
+        clusterSubscriptions.computeIfAbsent(clusterName, k -> new HashSet<>());
+    localSubscriptions.add(subscription);
+    addWatcher(new CdsWatcher(clusterName));
+
+    return subscription;
+  }
+
+  private boolean hasWatcher(XdsResourceType<?> type, String resourceName) {
+    TypeWatchers<?> typeWatchers = resourceWatchers.get(type);
+    return typeWatchers != null && typeWatchers.watchers.containsKey(resourceName);
+  }
+
+  @SuppressWarnings("unchecked")
+  private <T extends ResourceUpdate> void addWatcher(XdsWatcherBase<T> watcher) {
+    XdsResourceType<T> type = watcher.type;
+    String resourceName = watcher.resourceName;
+
+    this.syncContext.execute(() -> {
+      TypeWatchers<T> typeWatchers = (TypeWatchers<T>)resourceWatchers.get(type);
+      if (typeWatchers == null) {
+        typeWatchers = new TypeWatchers<>(type);
+        resourceWatchers.put(type, typeWatchers);
+      }
+
+      typeWatchers.add(resourceName, watcher);
+      xdsClient.watchXdsResource(type, resourceName, watcher);
```

Comment:
Use the 4-arg version, to pass in the executor to use (I assume `syncContext`). I think we would have deprecated/deleted this 3-arg version, but it is used a lot in tests. If you enter the syncContext manually within the callback, then ADS flow control stops working properly.

Reply:
You added `syncContext` here, but you _also_ added `syncContext.execute()` to the resource watchers. That means they never will do their processing in-line. Remove the `syncContext.execute()` from the resource watcher callbacks.